eats_worm/Extractor.py

```diff
@@ -467,9 +473,60 @@ def __init__(self, *args, **kwargs):
             pass
         else:
             kwargs['regen']=True
+            print(os.path.join(self.output_dir, 'params.json'))
             with open(os.path.join(self.output_dir, 'params.json'), 'w') as json_file:
                 json.dump(kwargs, json_file)
 
+    def process_im(self):
+        """Runs basic image pre-processing steps and saves processed image"""
+        if self.processing_params["neuroPAL"]:
+            channels = self.processing_params["RGBW_channels"]
+            NP_image = np.transpose(self.im.tf[0].asarray())
+            print(self.im.tf[0].asarray().shape)
+            NP_image = np.squeeze(NP_image[:,:,:,channels])
+            NP_image = NP_image.astype('int32')
+
+            print('Preprocessing image')
+
+            if self.processing_params["resample"]:   
+                print('Resampling image') 
+                new_res = self.processing_params["new_resolution"]
+                old_res = self.processing_params["old_resolution"]
+                NP_image = res.zoom_interpolate(new_res, old_res, NP_image)
+
+                print('Resampled image to: '+str(new_res))
+
+            if self.processing_params["median_filter"]:
+                print('Median filtering image')
+                size = self.processing_params.get("median", 3)
+
+                NP_image = medFilter2d(NP_image, size)
+
+            if self.processing_params["histogram_match"]:
+                # TODO: make sure reffiles are in tif format
+                print('Matching histogram of image to reference')
+
+                A_max = self.processing_params["A_max"]
+                ref_max = self.processing_params["ref_max"]
+
+                reffile = sio.loadmat(self.processing_params["im_to_match"])
+                refchannels = reffile['prefs']['RGBW'][0][0]-1
+                refdata = reffile['data']
+                refRGBW = np.squeeze(refdata[:,:,:,refchannels])
+
+                print(refRGBW.shape)
+
+                NP_image = hist.match_histogram(NP_image, refRGBW, A_max, ref_max)
+
+                print('Matched histogram of image to reference')
+
+            NP_image = NP_image.astype('uint16')
+            NP_image = np.transpose(NP_image, (2,3,1,0))
+
+            tifffile.imwrite(self.root+'/processed_data.tif', NP_image, imagej = True)
```

Comment:
I think I pretty much assumed that the use case with neuroPAL images would just be a single volume so would only be in a single tiff file. Handling of GCAMP time series data would be a totally separate process. So I have not really considered the multifile tiff case here, but let me know if that's something you think we should take into account

Reply:
Got it. I think that your intuition around separating preprocessing into its own routine is good, and I think that there's probably utility in expanding that routine to support fluorescence recordings in addition to NeuroPAL images, since it's cleaner to have the preprocessing steps self-contained and clearly defined than to handle them within each individual segmentation algorithm (which is what's currently happening). As above, though, we don't need to block these changes on that. Let's just make a comment note that the process call currently only supports NP images and open an issue in the tracker to make it work for fluorescence recordings as well